File: Fortran/IBM/execute/F2018/cintrop_ts29113/AssumedType/unit_tests/AssumedTypeObj05f.f90
!**********************************************************************
!*  ===================================================================
!*
!*  DATE                       : Apr 14, 2012
!*  ORIGIN                     : Linux/AIX Compiler Development,
!*                             : IBM Software Solutions China Lab
!*
!*  PRIMARY FUNCTIONS TESTED   : C-interop Assumed-type object
!*
!*  SECONDARY FUNCTIONS TESTED : None
!*
!*  REQUIRED COMPILER OPTIONS  :
!*
!*  DESCRIPTION                : Calling a non-BIND(C) procedure from Fortran
!*                               where the procedure is defined in C with
!*                               assumed-type object.
!*                               - Actual arg is of derived-type.
!*
!234567890123456789012345678901234567890123456789012345678901234567890

program AssumedTypeObj05f
  use, intrinsic :: iso_c_binding
  implicit none

  type c_intr1
    integer(c_int) :: a = 5
    integer(c_short) :: b = 10
    real(c_double) :: c = 1.2
  end type

  type(c_intr1) :: c_op

  call sub(c_op)

  c_op%a = 8
  c_op%b = 7
  c_op%c = 2.7

  call sub(c_op)

  contains
    subroutine sub(c_intr1)
      TYPE(*) :: c_intr1
      integer :: i

      interface
        integer function c_func_print_str(c_intr1)
          TYPE(*) :: c_intr1
        end function
      end interface

      i = c_func_print_str(c_intr1)

      if (i <> 15) then
        ERROR STOP 1
      end if
    end
end

File: Fortran/IBM/execute/F2018/cintrop_ts29113/AssumedType/unit_tests/AssumedTypeObj05f.c
/**********************************************************************
*  ===================================================================
*  XL Fortran Test Case                          IBM INTERNAL USE ONLY
*  ===================================================================
*
*  TEST CASE TITLE            : AssumedTypeObj05f
*
*  PROGRAMMER                 : Ren, Jian Gang
*  DATE                       : Apr 14, 2012
*  ORIGIN                     : Linux/AIX Compiler Development,
*                             : IBM Software Solutions China Lab
*
*  PRIMARY FUNCTIONS TESTED   : C-interop Assumed-type object
*                                                   
*  SECONDARY FUNCTIONS TESTED : None 
*
*  DRIVER STANZA              : xlc
*  REQUIRED COMPILER OPTIONS  : 
*
*
*  DESCRIPTION                : Calling a non-BIND(C) procedure from Fortran
*                               where the procedure is defined in C with
*                               assumed-type object.
*                               - Actual arg is of derived-type. 
*
*1234567890123456789012345678901234567890123456789012345678901234567890*/

#include <stdio.h>

typedef struct c_intr
{
  int a;
  short b;
  double c;  
}c_intr_t;

int c_func_print_str(void* a)
{
  c_intr_t *v = (struct c_intr *)a;

  printf("v->a: %d\n", v->a);
  printf("v->b: %d\n", v->b);
  printf("v->c: %g\n", v->c);
  
  return (v->a + v->b);
}
